### PR TITLE
MSC4213: Remove server_name parameter from `join` and `knock` endpoints

### PR DESCRIPTION
**Pull request by Johennes**
[Rendered](https://github.com/Johennes/matrix-spec-proposals/blob/johannes/remove-server-name/proposals/4213-remove-server-name.md)

---

In line with https://github.com/matrix-org/matrix-spec/issues/1700, the following disclosure applies:

I am a Systems Architect at [gematik](https://www.gematik.de/), Software Engineer at [Unomed](https://www.unomed.ch/), Matrix community member and former Element employee. This proposal was written and published with my community member hat on.

----

[FCP tickyboxes](https://github.com/matrix-org/matrix-spec-proposals/pull/4213#issuecomment-2484386461)

### CODE REVIEW

File: proposals/4213-remove-server-name.md
Comment (turt2live) [inline proposals/4213-remove-server-name.md:1]: Implementation requirements:

I'm tempted to mirror https://github.com/matrix-org/matrix-spec-proposals/pull/4127#discussion_r1560097122 and say "none", but am second-guessing that MSC's requirements too. It might be beneficial to have evidence that developers are picking up the newer approaches before suggesting removal from the spec.

Thoughts very welcome.

Comment (Johennes) [reply proposals/4213-remove-server-name.md:1]: It might not count because they were all created by me but there are five implementations linked in the proposal's body, four of which have already landed.

**Comment by turt2live** [comment]
Using the FCP mechanism to encourage comment/explicit blockers:

@mscbot fcp merge

**Comment by turt2live** [comment]
@mscbot concern implementation-needs-checking label

**Review by dbkr — COMMENTED** [review]
It's possible I'm confused but https://github.com/matrix-org/matrix-js-sdk/pull/4381 removes the unstable prefix from the 'via' param. This MSC is about removing the old `server_name` param, which is a different thing.

File: proposals/4213-remove-server-name.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+## Potential issues
+
+None. Servers can continue advertising support for earlier versions of the spec that included
+`server_name` via [`/_matrix/client/versions`].
+
+As of writing, the following stable implementations of [MSC4156] are known to the author:
+
+- synapse: https://github.com/element-hq/synapse/pull/17650
+- dendrite: https://github.com/matrix-org/dendrite/pull/3438
+- matrix-js-sdk: https://github.com/matrix-org/matrix-js-sdk/pull/4381
```
Comment (dbkr) [inline proposals/4213-remove-server-name.md:24]: It's possible I'm confused but https://github.com/matrix-org/matrix-js-sdk/pull/4381 removes the unstable prefix from the 'via' param. This MSC is about removing the old server_name param, which is a different thing.

Comment (tulir) [reply proposals/4213-remove-server-name.md:24]: The linked implementations are for MSC4156 which added the `via` param. They're listed to show that the old `server_name` parameter is safe to remove

Comment (Johennes) [reply proposals/4213-remove-server-name.md:24]: Yes, exactly. I think the `server_name` parameter will probably stay in implementations for quite a while unless they consciously choose to only support Matrix >= 1.12. But we don't need it in the spec anymore.

Comment (dbkr) [reply proposals/4213-remove-server-name.md:24]: Oh right, I see - I guess I was confused by your [comment](https://github.com/matrix-org/matrix-spec-proposals/pull/4213#discussion_r1794055534). In that case I guess I can vouch for the js-sdk impl, in that it now sends the via param.

File: proposals/4213-remove-server-name.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+[MSC4156] deprecated the `server_name` parameter on [`/_matrix/client/v3/join/{roomIdOrAlias}`]
+and [`/_matrix/client/v3/knock/{roomIdOrAlias}`] in favor of a new parameter `via`. This change
+shipped in [Matrix v1.12]. In line with the [deprecation policy], the `server_name` parameter
+is now eligible for removal from the spec.
+
+
+## Proposal
+
+The deprecated `server_name` parameter is removed from [`/_matrix/client/v3/join/{roomIdOrAlias}`]
+and [`/_matrix/client/v3/knock/{roomIdOrAlias}`].
```
Comment (clokep) [inline proposals/4213-remove-server-name.md:12]: I don't want to bikeshed too much, but it seems like clients need to look at `/versions` to decide what sort of parameters to give to a versioned endpoint...

tl;dr Should this rev v3 -> v4 for these endpoints since it is a backwards incompatible change?

Comment (richvdh) [reply proposals/4213-remove-server-name.md:12]: Can't the client just send both params? That's what EW seems to do, empirically.

Comment (Johennes) [reply proposals/4213-remove-server-name.md:12]: I suppose cutting a new version would be cleaner as it's a breaking change. But then I think clients would still have to check /versions to figure out which version of the endpoint they can call?

Comment (clokep) [reply proposals/4213-remove-server-name.md:12]: > Can't the client just send both params? That's what EW seems to do, empirically.

True -- I don't feel strongly, just wanted to bring it up as the endpoints *are* versioned for a reason!

Comment (KitsuneRal) [reply proposals/4213-remove-server-name.md:12]: The client sending both params for backcomp seems like a reasonable way forward, while servers are encouraged to ignore `via` in favour of `server_name`. We could just keep the endpoint the way it is right now (with the deprecated parameter), and the question would not even stand?

Comment (clokep) [reply proposals/4213-remove-server-name.md:12]: Seems folks are happy with this. 👍

**Comment by turt2live** [comment]
@mscbot resolve implementation-needs-checking label

File: proposals/4213-remove-server-name.md (hunk 2 above)
Comment (KitsuneRal) [reply proposals/4213-remove-server-name.md:12]: Sorry, looks like I was misunderstood. What I meant is that the API shape could remain as it is, with `server_name` deprecated; but homeservers would be expected to ignore it from Matrix 1.13 onwards. But it's not a hill for me to die on.

**Comment by Johennes** [comment]
Spec PR: https://github.com/matrix-org/matrix-spec/pull/2059

**Comment by turt2live** [comment]
Merged 🎉
